Add API Gateway endpoint configuration

alexdebrie opened the pull request:
## What did you implement:

Closes #4440.

Adds ability to specify [regional endpoints](https://aws.amazon.com/about-aws/whats-new/2017/11/amazon-api-gateway-supports-regional-api-endpoints/) for API Gateway. :

```yml
# serverless.yml

provider:
  name: aws
  endpointType: regional
...
```

## How did you implement it:

Adds the [EndpointConfiguration](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-restapi-endpointconfiguration.html) attribute when we compile the RestApi resource. By default, it sets to `EDGE`, which is the default API GW setting. 

## How can we verify it:

I used this example:

```yml
service: endpoint-test

# The `provider` block defines where your service will be deployed
provider:
  name: aws
  runtime: nodejs6.10
  endpointType: regional

functions:
  helloWorld:
    handler: handler.helloWorld
    events:
      - http:
          path: hello-world
          method: get
          cors: true
```

Then run `sls package` and check the `.serverless/cloudformation-template-update-stack.json` file that the RestAPI resource has the EndpointConfiguration defined as desired.

## Todos:

- [X] Write tests
- [X] Write documentation
- [X] Fix linting errors
- [ ] Make sure code coverage hasn't dropped
- [ ] Provide verification config / commands / resources
- [ ] Enable "Allow edits from maintainers" for this PR
- [ ] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

HyperBrain commented on lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js:
```diff
@@ -7,11 +7,31 @@ module.exports = {
+    let endpointType = 'EDGE';
+
+    if (this.serverless.service.provider.endpointType) {
+      const validEndpointTypes = ['REGIONAL', 'EDGE'];
```
According to the AWS docs, an endpoint type can also be a domain name. Should we support this? I think it could be important as soon as you map a custom domain names to your API - but I did not test anything in that area yet.

alexdebrie replied:
@HyperBrain I think it's slightly different than that. A [domain name that you register with API Gateway](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-domainname.html#cfn-apigateway-domainname-endpointconfiguration) can also have an endpoint configuration of Edge / Regional. Thus, if you set up `api.mycompany.com` in API Gateway, you could configure it to be Regional.

Currently, the Framework doesn't configure API GW domains in core, so I don't think we can do anything with that. It could be a feature for the [serverless-domain-manager](https://github.com/amplify-education/serverless-domain-manager) plugin to add.

I'm not 100% sure I'm correct on this so feel free to correct me :)

HyperBrain replied:
The reason why I came to the assumption that it could be defined here is, that the CF AWS::ApiGateway::RestApi resource defines its `EndpointConfiguration` property here in that way: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-restapi-endpointconfiguration.html

There is written "... or its custom domain name". Additionally it seems that the property (which is an array) can contain multiple values. But don't ask me how exactly this is intended to work or when exactly you should add multiple values to the array.

So this is for me a bit unclear too 😄

alexdebrie replied:
Good point. That's confusing.

For what it's worth, the [EndpointConfiguration](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-restapi-endpointconfiguration.html) documentation for the RestApi resource is nearly identical to the [EndpointConfiguration](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-domainname-endpointconfiguration.html) for the DomainName resource. I wonder if that sentence -- `A list of endpoint types of an API or its custom domain name` refers to the fact that this configuration is similar for both a RestApi resource and a DomainName resource?

Agreed that the array makes no sense 😄

HyperBrain commented:
@alexdebrie Just saw that there are still ESLint failures:
```
/home/travis/build/serverless/serverless/lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.test.js
  19:13  error  Strings must use singlequote             quotes
  19:19  error  Missing trailing comma                   comma-dangle
  62:70  error  Strings must use singlequote             quotes
  71:1   error  Block must not be padded by blank lines  padded-blocks
```

alexdebrie commented:
@HyperBrain: 

> Imo the 2-value solution will get users onto the feature, so that the rest will happen automatically

That's a great point. Sort of a paraphrase Linus's Law -- "Given enough eyeballs, all ~~bugs~~ _bad AWS documention_ is shallow. 😄

I fixed the linting errors -- my mistake for not grabbing those initially. I'm still not sure why code coverage it thought code coverage was dropping, but it fixed now.

@horike37 

Great catch on that and big mistake on my part. I missed that `EndpointConfiguration` has a `Types` attribute where you list the endpoint types. I could have sworn I tested it with a real deploy. Either way, I have fixed it now. Please let me know if it works for you.

horike37 commented:
Thank you for the update @alexdebrie  :+1: 
just tested it again, and confirmed working fine as expected.

Seems everything is OK. merging...

```
$aws apigateway get-rest-api --rest-api-id j75f4zgmzb
{
    "name": "dev-testesttestss", 
    "endpointConfiguration": {
        "types": [
            "REGIONAL"
        ]
    }, 
    "id": "j75f4zgmzb", 
    "createdDate": 1512918617
}

```
